src/slurm_cloud_bursting/resume.py

```diff
@@ -0,0 +1,262 @@
+
+        log.info(self.__repr__())
+
+
+@retry(stop_max_attempt_number=3, wait_fixed=5000)
+def _set_nodes_down_and_power_save(node_list):
+    set_nodes_down(node_list, reason="Failure when resuming nodes")
+    set_nodes_power_down(node_list, reason="Failure when resuming nodes")
+
+
+def _handle_failed_nodes(node_list):
+    """
+    Fall back mechanism to handle failure when launching instances.
+
+    When encountering a failure, want slurm to deallocate current nodes,
+    and re-queue job to be run automatically by new nodes.
+    To do this, set node to DOWN, so slurm will automatically re-queue job.
+    Then set node to POWER_DOWN so suspend program will be run.
+    Suspend program needs to properly clean up instances(if any) and set node back to IDLE in all cases.
+
+    If this process is not done explicitly, slurm will wait until ResumeTimeout,
+    then execute this process of setting nodes to DOWN then POWER_DOWN.
+    To save time, should explicitly set nodes to DOWN then POWER_DOWN after encountering failure.
+    """
+    try:
+        log.info("Node %s marked as down and placed into power_down", node_list)
+        _set_nodes_down_and_power_save(node_list)
+    except Exception as e:
+        log.exception("Failed to place nodes %s into down/power_down with exception: %s", node_list, e)
+
+
+def _parse_ec2_instance(instance):
+    """Parse and return EC2 instance info."""
+    return EC2Instance(
+        instance["InstanceId"],
+        private_ip=instance["PrivateIpAddress"],
+        hostname=instance["PrivateDnsName"].split(".")[0],
+    )
+
+
+def _update_slurm_node_addrs(launched_nodes):
+    """Update node information in slurm with info from launched EC2 instance."""
+    for instance, slurm_node in launched_nodes:
+        try:
+            update_nodes(slurm_node, nodeaddr=instance.private_ip, nodehostname=instance.hostname, raise_on_error=True)
+            log.info(
+                "Node %s is now configured with instance=%s private_ip=%s nodehostname=%s",
+                slurm_node,
+                instance.id,
+                instance.private_ip,
+                instance.hostname,
+            )
+        except subprocess.CalledProcessError:
+            log.error(
+                "Encountered error when updating node %s with instance=%s private_ip=%s nodehostname=%s",
+                slurm_node,
+                instance.id,
+                instance.private_ip,
+                instance.hostname,
+            )
+            failed_nodes.append(slurm_node)
+
+
+def _update_run_instance_args(run_instances_args, cluster_name, queue, instance_type):
+    # LaunchTemplate is different for every instance type in every queue
+    # LaunchTemplate name format: {cluster_name}-{queue_name}-{instance_type}
+    # Sample LT name: hit-queue1-c5.xlarge
+    run_instances_args["LaunchTemplate"] = {"LaunchTemplateName": f"{cluster_name}-{queue}-{instance_type}"}
+    run_instances_args["InstanceType"] = instance_type
```

Comment:
Since now we have a separate LT for each instance type I think we only need to select the right LT, so probably we can get rid of this  function

Reply:
Ah ok